File: src/data/courses/dsa/trees/path-relationship-problems/max-path-sum/code/python/max_path_sum.py
# ============================================================
# Binary Tree Maximum Path Sum
# ============================================================
class Solution:
    def maxPathSum(self, root) -> int:
        self.global_max = float('-inf')

        def max_gain(node):
            if not node: return 0
            left = max(0, max_gain(node.left))   # Ignore negative branches
            right = max(0, max_gain(node.right))
            # Path through this node (uses both sides)
            self.global_max = max(self.global_max, node.val + left + right)
            # Return to parent: can only extend ONE side
            return node.val + max(left, right)

        max_gain(root)
        return self.global_max

File: src/data/courses/dsa/trees/path-relationship-problems/max-path-sum/code/java/MaxPathSum.java
// ============================================================
// Binary Tree Maximum Path Sum
// ============================================================
class Solution {
    int globalMax = Integer.MIN_VALUE;
    public int maxPathSum(TreeNode root) {
        globalMax = Integer.MIN_VALUE;
        maxGain(root);
        return globalMax;
    }
    int maxGain(TreeNode node) {
        if (node == null) return 0;
        int left = Math.max(0, maxGain(node.left));
        int right = Math.max(0, maxGain(node.right));
        globalMax = Math.max(globalMax, node.val + left + right);
        return node.val + Math.max(left, right);
    }
}
